src/main/java/org/opensearch/knn/index/codec/jvector/JVectorGraphIndexWriter.java

```diff
@@ -0,0 +1,277 @@
+
+public class JVectorGraphIndexWriter implements AutoCloseable {
+    private final IndexOutput outData;
+    private final IndexOutput outMeta;
+    private final int version;
+    private final long startOffset;
+    private final GraphIndex graph;
+    private final OrdinalMapper ordinalMapper;
+    private final int dimension;
+    private final GraphIndex.View view;
+    private final EnumMap<FeatureId, Feature> featureMap;
+    private final int headerSize;
+    private final List<Feature> inlineFeatures;
+
+    public JVectorGraphIndexWriter(
+        IndexOutput outData,
+        IndexOutput outMeta,
+        int version,
+        long startOffset,
+        GraphIndex graph,
+        OrdinalMapper oldToNewOrdinals,
+        int dimension,
+        EnumMap<FeatureId, Feature> features
+    ) {
+        assert outData != null : "RandomAccessWriter cannot be null";
+        assert version >= 0 : "Version cannot be negative";
+        assert startOffset >= 0 : "Start offset cannot be negative";
+        assert graph != null : "Graph cannot be null";
+        assert oldToNewOrdinals != null : "Ordinal mapper cannot be null";
+        assert dimension > 0 : "Dimension must be greater than 0";
+        assert features != null : "Features cannot be null";
+        assert !features.isEmpty() : "Features cannot be empty";
+        this.outData = outData;
+        this.outMeta = outMeta;
+        this.version = version;
+        this.startOffset = startOffset;
+        this.graph = graph;
+        this.ordinalMapper = oldToNewOrdinals;
+        this.dimension = dimension;
+        this.view = graph instanceof OnHeapGraphIndex ? ((OnHeapGraphIndex) graph).getFrozenView() : graph.getView();
+        this.featureMap = features;
+        this.inlineFeatures = features.values().stream().filter(f -> !(f instanceof SeparatedFeature)).collect(Collectors.toList());
+
+        // create a mock Header to determine the correct size
+        var layerInfo = CommonHeader.LayerInfo.fromGraph(graph, ordinalMapper);
+        var ch = new CommonHeader(version, dimension, 0, layerInfo, 0);
+        var placeholderHeader = new Header(ch, featureMap);
+        this.headerSize = placeholderHeader.size();
+    }
+
+    public static JVectorGraphIndexWriter create(
+        IndexOutput outData,
+        IndexOutput outMeta,
+        long startOffset,
+        GraphIndex graphIndex,
+        int dimension,
+        EnumMap<FeatureId, Feature> features
+    ) {
+        var version = OnDiskGraphIndex.CURRENT_VERSION;
+        var ordinalMapper = new OrdinalMapper.MapMapper(sequentialRenumbering(graphIndex));
+        return new JVectorGraphIndexWriter(outData, outMeta, version, startOffset, graphIndex, ordinalMapper, dimension, features);
+    }
+
+    // @Override
+    public synchronized Header getHeader() throws IOException {
+        // graph-level properties
+        // out.seek(startOffset);
```

Comment:
I left the commented code on purpose so it's easy to inspect the changes and how we no longer do any random repositioning of the write (seek)